fvm/transactionStorageLimiter_test.go

```diff
@@ -0,0 +1,201 @@
+		owner := string(flow.HexToAddress("1").Bytes())
+		ledger := newMockLedger(
+			[]string{owner, owner},
+			[]OwnerKeyValue{
+				storageUsed(owner, 99),
+				storageCapacity(owner, 100),
+				accountExists(owner),
+			})
+		d := &fvm.TransactionStorageLimiter{}
+
+		err := d.Process(nil, fvm.Context{}, nil, ledger)
+
+		require.NoError(t, err)
+		// three touches per account: get exists, get capacity, get used
+		require.Equal(t, 3, ledger.GetCalls[owner])
+	})
+	t.Run("two registers change on different accounts, only check capacity once per account", func(t *testing.T) {
+		owner1 := string(flow.HexToAddress("1").Bytes())
+		owner2 := string(flow.HexToAddress("2").Bytes())
+		ledger := newMockLedger(
+			[]string{owner1, owner1, owner2, owner2},
+			[]OwnerKeyValue{
+				storageUsed(owner1, 99),
+				storageCapacity(owner1, 100),
+				accountExists(owner2),
+				storageUsed(owner2, 999),
+				storageCapacity(owner2, 1000),
+				accountExists(owner2),
+			})
+		d := &fvm.TransactionStorageLimiter{}
+
+		err := d.Process(nil, fvm.Context{}, nil, ledger)
+
+		require.NoError(t, err)
+		// three touches per account: get exists, get capacity, get used
+		require.Equal(t, 3, ledger.GetCalls[owner1])
+		require.Equal(t, 3, ledger.GetCalls[owner2])
+	})
+	t.Run("non account registers are ignored", func(t *testing.T) {
+		owner := ""
+		ledger := newMockLedger(
+			[]string{owner},
+			[]OwnerKeyValue{
+				storageUsed(owner, 101),
+				storageCapacity(owner, 100),
+				accountExists(owner), // it has exists value, but it cannot be parsed as an address
+			})
+		d := &fvm.TransactionStorageLimiter{}
+
+		err := d.Process(nil, fvm.Context{}, nil, ledger)
+
+		require.NoError(t, err)
+	})
+	t.Run("account registers without exists are ignored", func(t *testing.T) {
+		owner := string(flow.HexToAddress("1").Bytes())
+		ledger := newMockLedger(
+			[]string{owner},
+			[]OwnerKeyValue{
+				storageUsed(owner, 101),
+				storageCapacity(owner, 100),
+			})
+		d := &fvm.TransactionStorageLimiter{}
+
+		err := d.Process(nil, fvm.Context{}, nil, ledger)
+
+		require.NoError(t, err)
+	})
+}
+
+type MockLedger struct {
+	UpdatedRegisterKeys []flow.RegisterID
+	StorageValues       map[string]map[string]flow.RegisterValue
+	GetCalls            map[string]int
+}
+
+type OwnerKeyValue struct {
+	Owner string
+	Key   string
+	Value uint64
+}
+
+func storageUsed(owner string, value uint64) OwnerKeyValue {
+	return OwnerKeyValue{
+		Owner: owner,
+		Key:   "storage_used",
+		Value: value,
+	}
+}
+
+func storageCapacity(owner string, value uint64) OwnerKeyValue {
+	return OwnerKeyValue{
+		Owner: owner,
+		Key:   "storage_capacity",
+		Value: value,
+	}
+}
+
+func accountExists(owner string) OwnerKeyValue {
+	return OwnerKeyValue{
+		Owner: owner,
+		Key:   "exists",
+		Value: 1,
+	}
+}
+
+func newMockLedger(updatedKeys []string, ownerKeyStorageValue []OwnerKeyValue) MockLedger {
```

Comment:
Would it be possible to use `MapLedger` in place of this mock? I think you could keep this helper function to seed the ledger state -- I'm just not sure you need to define an extra data structure

Reply:
I wanted to check how many times I access the ledger get function. I don't think I can do that part with the MapLedger. I could remove those tests, since they are only testing that storage capacity checking is done once per account.